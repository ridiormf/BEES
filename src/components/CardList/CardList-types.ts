import { BulletProps } from '../Bullet/Bullet-types';

export interface CardListBulletProps extends BulletProps {
  children: string;
}

export interface CardListDataItem {
  key: string | number;
  title: string | number;
  firstLineText: string | number;
  firstLineEmptyText: string | number;
  secondLineText: string | number;
  secondLineEmptyText: string | number;
  bullets?: Array<CardListBulletProps>;
}

export interface CardListProps {
  data: Array<CardListDataItem>;
  bulletsWithAddMore?: boolean;
  onSaveNewBullet?: (cardId: string | number, value: string) => void;
  onDeleteCard?: (cardId: string | number, index?: number) => void;
}
