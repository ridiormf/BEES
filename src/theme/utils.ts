type Space = number | string;

const utils = {
  marginHorizontal: (value: Space) => ({
    marginLeft: value,
    marginRight: value,
  }),
  marginVertical: (value: Space) => ({
    marginTop: value,
    marginBottom: value,
  }),
  paddingHorizontal: (value: Space) => ({
    paddingLeft: value,
    paddingRight: value,
  }),
  paddingVertical: (value: Space) => ({
    paddingTop: value,
    paddingBottom: value,
  }),
};

export default utils;
