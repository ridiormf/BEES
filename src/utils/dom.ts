export const getBody = () => {
  return document.getElementsByTagName('body')[0];
};
