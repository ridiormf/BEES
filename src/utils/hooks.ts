import React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';

/**
 * Function that simulates the behavior of a class component constructor,
 * adapted to hooks, for functional components.
 *
 * @param {function} callback Callback which will be called as soon as the component is instantiated.
 */
export const useConstructor = <ValueType = void>(
  callback: () => ValueType,
): ValueType | void => {
  const componentWasBuilt = React.useRef<boolean>(false);
  const value = React.useRef<ValueType>();
  if (!componentWasBuilt.current) {
    componentWasBuilt.current = true;
    value.current = callback();
  }

  return value.current;
};

/**
 * Function that simulates the behavior of a class component method "componentDidMount",
 * adapted to hooks, for functional components.
 *
 * @param {function} callback Callback which will be called as soon as the component mount.
 */
export const useDidMount = (callback: () => void): void => {
  React.useEffect(callback, []);
};

/**
 * Function that simulates the behavior of a class component method "componentWillUnmount",
 * adapted to hooks, for functional components.
 *
 * @param {function} callback Callback which will be called as soon as the component will unmount.
 */
export const useWillUnmount = (callback: () => void): void => {
  React.useEffect(() => {
    return callback;
  }, []);
};

/**
 * Function that simulates the behavior of a class component method "componentDidUpdate",
 * adapted to hooks, for functional components.
 *
 * @param {function} callback Callback which will be called on each value change of variables declared as dependencies,
 * excluding the first render of the component.
 * @param {React.DependencyList} deps Dependency list to be listened.
 */
export const useDidUpdate = (
  callback: () => void,
  deps: React.DependencyList,
): void => {
  const firstRender = React.useRef<boolean>(true);
  React.useEffect(() => {
    if (firstRender.current) {
      firstRender.current = false;
    } else {
      callback();
    }
  }, deps);
};

/**
 * Function that use main hooks from "react-router-dom" to simplify usability.
 *
 * @returns {object} containing following attributes:
 * @type {function} "goBack": to return navigation one page.
 * @type {function} "go": to navigate for path.
 * @type {string} "pathname": current url pathname.
 */
export const useAppNavigation = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const goBack = () => {
    navigate(-1);
  };

  const go = (path: string) => {
    navigate(path);
  };

  return {
    goBack,
    go,
    pathname: location.pathname,
  };
};
