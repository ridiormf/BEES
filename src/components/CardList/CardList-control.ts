import React from 'react';
import { CardListProps } from './CardList-types';

export const useCardListControl = ({
  onSaveNewBullet,
}: Partial<CardListProps>) => {
  const [showBulletField, setShowBulletField] = React.useState<string>();
  const [fieldValue, setFieldValue] = React.useState<string>('');

  const openBulletField = (cardId: string) => {
    setFieldValue('');
    setShowBulletField(cardId);
  };

  const saveNewBullet = () => {
    if (onSaveNewBullet && showBulletField) {
      onSaveNewBullet(showBulletField, fieldValue);
    }
    setFieldValue('');
    setShowBulletField(undefined);
  };

  const onFieldKeyUp = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') saveNewBullet();
  };

  const onChangeField = (event: React.ChangeEvent<HTMLInputElement>) => {
    setFieldValue(event.target.value);
  };

  return {
    state: {
      showBulletField,
      fieldValue,
    },
    methods: {
      openBulletField,
      onChangeField,
      saveNewBullet,
      onFieldKeyUp,
    },
  };
};
