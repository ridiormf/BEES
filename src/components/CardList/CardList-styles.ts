import { CSS } from '@stitches/react';
import theme from '../../theme';
import ComponentText from '../Text';

const CardsContainer = theme.styled('div', {
  display: 'grid',
  gridTemplateColumns: 'repeat(1, 1fr)',
  gap: 40,
  alignItems: 'stretch',
  justifyItems: 'center',
  alignSelf: 'center',
  width: '100%',
  overflow: 'hidden',
  maxWidth: 1280,
  '@lgUp': {
    gridTemplateColumns: 'repeat(2, 1fr)',
  },
  '@xlUp': {
    gridTemplateColumns: 'repeat(3, 1fr)',
  },
  '@mdDown': {},
  // flexDirection: 'row',
  // flexWrap: 'wrap',
  // alignItems: 'center',
  // justifyContent: 'space-around',
});

const Card = theme.styled('div', {
  display: 'flex',
  flexDirection: 'column',
  position: 'relative',
  paddingHorizontal: 23,
  paddingVertical: 30,
  width: '100%',
  minWidth: 0,
  backgroundColor: '$secondaryCard',
  borderRadius: '$borderRadius',
  border: '1px solid $primaryText',
});

const CloseButton = theme.styled('button', {
  padding: 12,
  lineHeight: 0,
  fontSize: 16,
  position: 'absolute',
  top: 0,
  right: 0,
  border: 'none',
  backgroundColor: 'unset',
  cursor: 'pointer',
  transition: '200ms',
  alignSelf: 'flex-end',
  '&:hover': {
    filter: 'brightness(1.5)',
  },
  '&:active': {
    filter: 'brightness(2)',
  },
});

const Title = theme.styled('h3', {
  fontSize: 20,
  lineHeight: '30px',
  margin: 0,
  fontWeight: 700,
  color: '$secondaryTitle',
  marginBottom: 16,
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
});

const Text = theme.styled(ComponentText, {
  fontSize: 16,
  lineHeight: '26px',
  color: '$secondaryText',
  marginVertical: 4,
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
});

const BulletsContainer = theme.styled('div', {
  display: 'flex',
  flexDirection: 'row',
  gap: 16,
  alignItems: 'center',
  flexWrap: 'wrap',
});

const Bullet = theme.styled('button', {
  paddingVertical: 0,
  paddingHorizontal: 16,
  lineHeight: 0,
  fontSize: 16,
  minHeight: 38,
  border: 'none',
  borderRadius: 200,
  backgroundColor: '$primaryBackground',
  cursor: 'pointer',
  transition: '200ms',
  display: 'flex',
  alignItems: 'center',
  '&:hover': {
    filter: 'brightness(1.05)',
  },
  '&:active': {
    filter: 'brightness(1.1)',
  },
});

const TextEmptyCSS: CSS = {
  opacity: 0.5,
  fontStyle: 'italic',
  textTransform: 'lowercase',
};

export const cardListStyles = {
  styled: {
    CardsContainer,
    Card,
    CloseButton,
    Title,
    Text,
    BulletsContainer,
    Bullet,
  },
  css: {
    TextEmpty: TextEmptyCSS,
  },
};
