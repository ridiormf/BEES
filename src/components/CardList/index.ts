import CardList from './CardList';
import type { CardListProps, CardListDataItem } from './CardList-types';

export default CardList;
export type { CardListProps, CardListDataItem };
