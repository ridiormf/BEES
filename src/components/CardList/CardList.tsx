import React from 'react';
import { cardListStyles } from './CardList-styles';
import { CardListProps, DataItem } from './CardList-types';

import Image from '../Image';
import Bullet from '../Bullet';

import garbageSvg from '../../assets/svg/trash.svg';
import addMoreSvg from '../../assets/svg/plus-circle.svg';

const {
  styled: { CardsContainer, Card, CloseButton, Title, Text, BulletsContainer },
  css,
} = cardListStyles;

const CardList: React.FC<CardListProps> = ({ data, bulletsWithAddMore }) => {
  return (
    <CardsContainer>
      {data.map((item: DataItem) => (
        <Card key={item.key}>
          <CloseButton>
            <Image src={garbageSvg} />
          </CloseButton>
          <Title>{item.title}</Title>
          <Text css={item.firstLineText ? undefined : css.TextEmpty}>
            {item.firstLineText ? item.firstLineText : item.firstLineEmptyText}
          </Text>
          <Text css={item.secondLineText ? undefined : css.TextEmpty}>
            {item.secondLineText
              ? item.secondLineText
              : item.secondLineEmptyText}
          </Text>
          {item.bullets ? (
            <BulletsContainer>
              {item.bullets.map((bullet) => (
                <Bullet icon={bullet.icon} key={bullet.children}>
                  {bullet.children}
                </Bullet>
              ))}
              {bulletsWithAddMore ? (
                <Bullet icon={addMoreSvg} onClick={() => {}}>
                  add more
                </Bullet>
              ) : undefined}
            </BulletsContainer>
          ) : null}
        </Card>
      ))}
    </CardsContainer>
  );
};

export default CardList;
