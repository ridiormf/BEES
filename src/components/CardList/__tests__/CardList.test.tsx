import { render } from '@testing-library/react';
import { Wrapper } from '../../../__tests__/utils';
import CardList from '../CardList';

describe('components ->', () => {
  describe('CardList ->', () => {
    describe('snapshots ->', () => {
      it('Should match snapshot', () => {
        const { container } = render(
          <CardList bulletsWithAddMore data={[]} />,
          {
            wrapper: Wrapper,
          },
        );
        expect(container).toMatchSnapshot();
      });
    });
  });
});
