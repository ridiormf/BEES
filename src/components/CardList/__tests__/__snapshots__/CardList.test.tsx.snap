// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components -> CardList -> snapshots -> Should match snapshot 1`] = `
<div>
  <div
    class="c-hnDHzq"
  />
</div>
`;
